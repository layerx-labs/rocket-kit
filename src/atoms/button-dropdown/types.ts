export interface ActionMenu<T> {
  id: string;
  value: string;
  visibilityFunc?: (data: T) => boolean;
  url?: string;
  action?: (
    e: React.MouseEvent<HTMLAnchorElement, MouseEvent> | React.ChangeEvent<HTMLSelectElement>,
    data: T | null
  ) => void;
}
