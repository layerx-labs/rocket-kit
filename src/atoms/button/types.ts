export type ButtonVariant = 'solid' | 'outline' | 'text';

export type ButtonColor =
  | 'primary'
  | 'danger'
  | 'info'
  | 'purple'
  | 'white'
  | 'black'
  | 'magic'
  | 'pulse';
