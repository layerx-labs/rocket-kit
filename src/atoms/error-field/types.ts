export type ErrorFieldColor = 'success' | 'red';
