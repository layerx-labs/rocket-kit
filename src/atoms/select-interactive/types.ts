import { MapHTMLAttributes, ReactNode } from 'react';
import NamedProps from 'react-select/dist/declarations/src/Select';

export interface TSelectInteractiveOption {
  value: string;
  label: string;
  [k: string]: any;
}

export interface TOptions extends TSelectInteractiveOption {
  icon?: string;
  customImage?: string;
}

export interface TCustomOptions {
  data: TOptions;
}

export interface OptionsGroup<T extends TSelectInteractiveOption> {
  label: string;
  options: ReadonlyArray<T>;
}

type TDivElement = Omit<
  MapHTMLAttributes<HTMLDivElement>,
  'onChange' | 'placeholder'
>;

type Options =
  | TSelectInteractiveOption
  | ReadonlyArray<TSelectInteractiveOption>;

export interface SelectInteractiveProps<T extends Options>
  extends TDivElement,
    Partial<Pick<NamedProps<T>, 'onInputChange'>> {
  name?: string;
  error?: string;
  clear?: boolean;
  multi?: boolean;
  search?: boolean;
  disabled?: boolean;
  placeholder?: ReactNode;
  formatGroupLabel?: boolean;
  value?: ReadonlyArray<TOptions> | TOptions | null | undefined;
  onChange: (values: Array<TOptions> | TOptions) => void;
  options: ReadonlyArray<TOptions | OptionsGroup<TOptions>>;
}
