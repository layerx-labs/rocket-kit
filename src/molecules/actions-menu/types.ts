export type ActionMenuType = 'danger' | 'default';

export interface ActionMenu<T> {
  id: string;
  type: ActionMenuType;
  value: string;
  visibilityFunc?: (data: T) => boolean;
  url?: string;
  action?: (
    e: React.MouseEvent<HTMLAnchorElement, MouseEvent>,
    data: T | null
  ) => void;
}
