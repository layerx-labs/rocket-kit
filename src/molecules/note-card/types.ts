export type NoteColor = 'grey' | 'green' | 'red';
