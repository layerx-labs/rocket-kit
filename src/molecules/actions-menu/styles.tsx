import styled from 'styled-components';
import { lighten } from 'polished';
import { colors } from '../../ions/variables';

const { normal, light, info, danger } = colors;

export const ActionsMenuStyle = styled.div`
  height: 50px;

  button {
    margin-top: 7px;
  }

  ul {
    position: absolute;
    border: 1px solid ${lighten(0.4, info)};
    border-radius: 4px;
    background-color: ${light};
    margin-top: 5px;
    min-width: 200px;
    max-width: 250px;
    padding: 0;
    -moz-box-shadow: 0 0 15px 0 ${lighten(0.7, normal)};
    -webkit-box-shadow: 0 0 15px 0 ${lighten(0.7, normal)};
    box-shadow: 0 0 15px 0 ${lighten(0.7, normal)};
    z-index: 1;

    li {
      list-style: none;
      min-height: 45px;
      display: flex;
      transition-duration: 0.3s;

      &:hover {
        background-color: ${lighten(0.4, info)};
        cursor: pointer;
      }

      &:first-child {
        border-radius: 4px 4px 0 0;
      }

      &:last-child {
        border-radius: 0 0 4px 4px;
      }

      &.danger {
        border-top: 1px solid ${lighten(0.4, info)};

        a {
          color: ${danger};
        }
      }

      a {
        width: 100%;
        display: flex;
        align-items: center;
        padding: 0 20px;
        color: ${normal};
        text-decoration: none;
        white-space: nowrap;

        span {
          display: table-cell;
          overflow: hidden;
          text-overflow: ellipsis;
        }
      }
    }
  }
`;
