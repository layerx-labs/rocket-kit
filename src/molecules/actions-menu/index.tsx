import React, { CSSProperties } from 'react';
import * as Styles from './styles';
import useVisible from '../../utils/hooks/use-visible';
import { Button } from '../..';
import { ActionMenu } from './types';

interface ActionsMenuInterface<T> {
  className?: string;
  style?: CSSProperties;
  ariaLabel?: string;
  actions: ActionMenu<T>[];
  data?: T;
  dataTestId?: string;
  startsOpen?: boolean;
}

const ActionsMenu = <T,>(props: ActionsMenuInterface<T>) => {
  const {
    ariaLabel = '',
    actions = [],
    data = null,
    dataTestId = 'action-menu-button',
    startsOpen = false,
  } = props;

  const { ref, isVisible, setIsVisible } = useVisible<HTMLDivElement>(
    startsOpen
  );

  if (!actions || !Array.isArray(actions) || actions.length === 0) return <></>;

  const visibleActions = actions.filter(({ visibilityFunc = null }) =>
    data && visibilityFunc ? visibilityFunc(data) : true
  );

  if (visibleActions.length === 0) return <></>;

  return (
    <Styles.ActionsMenuStyle ref={ref}>
      <Button
        variant="text"
        color="dark"
        icon="menuVert"
        ariaLabel={ariaLabel}
        action={evt => {
          evt.preventDefault();
          setIsVisible(!isVisible);
          evt.stopPropagation();
        }}
        dataTestId={dataTestId}
      />

      {isVisible ? (
        <ul data-testid={'ul-action-menu'}>
          {visibleActions.map(
            ({
              id = '',
              type = '',
              url = '',
              action = () => {},
              value = '',
            }) => (
              <li
                key={id}
                data-testid={'li-action-menu'}
                className={type === 'danger' ? 'danger' : undefined}
              >
                <a
                  href={url}
                  onClick={e => {
                    setIsVisible(false);
                    action(e, data);
                    e.stopPropagation();
                  }}
                >
                  <span>{value}</span>
                </a>
              </li>
            )
          )}
        </ul>
      ) : null}
    </Styles.ActionsMenuStyle>
  );
};

export default ActionsMenu;
