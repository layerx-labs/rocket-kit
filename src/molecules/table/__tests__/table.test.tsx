import React from 'react';
import { render, screen } from '@testing-library/react';
import { ActionMenu } from '../../actions-menu/types';
import { Avatar, Table } from '../../..';
import userEvent from '@testing-library/user-event';

describe('Table', () => {
  it('renders', async () => {
    interface CellData {
      id: string;
      transactionName: string;
      amount: string;
      type: string;
      createdAt: string;
      avatar: string;
    }

    const onClick = jest.fn();
    const actionText = 'Go to Details';
    const actions: ActionMenu<CellData>[] = [
      {
        id: 'actionDetails',
        type: 'default',
        value: actionText,
        action: onClick,
      },
    ];

    const rows: CellData[] = [
      {
        id: '1',
        transactionName: 'To the Moon',
        amount: '2700',
        type: 'Back',
        createdAt: '2018-07-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
    ];

    const options = {
      columns: [
        {
          id: 'transactionName',
          value: 'Last Transaction',
          dataKey: 'transactionName',
          renderer: (transactionName: string, cell: CellData) => (
            <>
              <Avatar url={cell.avatar} alt={transactionName} />{' '}
              {transactionName}
            </>
          ),
          className: 'avatar',
        },
        {
          id: 'amount',
          value: 'Amount',
          dataKey: 'amount',
          className: 'kai',
        },
      ],
    };

    const { asFragment } = render(
      <Table<CellData> options={options} values={rows} actions={actions} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders 5 rows', async () => {
    interface CellData {
      id: string;
      transactionName: string;
      amount: string;
      type: string;
      createdAt: string;
      avatar: string;
    }

    const rows: CellData[] = [
      {
        id: '1',
        transactionName: 'To the Moon 1',
        amount: '2700',
        type: 'Back',
        createdAt: '2018-07-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
      {
        id: '2',
        transactionName: 'To the Moon 2',
        amount: '270000',
        type: 'Back',
        createdAt: '2019-07-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
      {
        id: '3',
        transactionName: 'To the Moon 3',
        amount: '99700',
        type: 'Back',
        createdAt: '2019-08-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
      {
        id: '4',
        transactionName: 'To the Moon 4',
        amount: '700',
        type: 'Back',
        createdAt: '2019-09-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
      {
        id: '5',
        transactionName: 'To the Moon 5',
        amount: '2',
        type: 'Back',
        createdAt: '2020-07-7 16:21:13',
        avatar: './images/default-avatar.svg',
      },
    ];

    const options = {
      columns: [
        {
          id: 'amount',
          value: 'Amount',
          dataKey: 'amount',
          className: 'kai',
          dataTestId: 'header-amount-id',
        },
        {
          id: 'type',
          value: 'Type',
          dataKey: 'type',
          className: 'center',
        },
        {
          id: 'created',
          value: 'Created',
          dataKey: 'createdAt',
          className: 'createdAt',
        },
        {
          id: 'transaction',
          value: 'Transaction',
          dataKey: 'transactionName',
        },
      ],
    };

    render(<Table<CellData> options={options} values={rows} />);
    for (let row = 1; row < 6; ++row) {
      screen.getByText(`To the Moon ${row}`);
    }
    expect(screen.queryAllByTestId(`td-header-amount-id`)).toHaveLength(5);
  });

  it('renders all table headers correctly', async () => {
    interface CellData {
      id: string;
      amount: string;
      type: string;
      createdAt: string;
    }

    const rows: CellData[] = [
      {
        id: '1',
        amount: '2700',
        type: 'Back',
        createdAt: '2018-07-7 16:21:13',
      },
    ];

    const headerAmount = 'Amount';
    const headerType = 'Type';
    const headerCreated = 'Created';
    const options = {
      columns: [
        {
          id: 'amount',
          value: headerAmount,
          dataKey: 'amount',
          className: 'kai',
        },
        {
          id: 'type',
          value: headerType,
          dataKey: 'type',
          className: 'center',
        },
        {
          id: 'created',
          value: headerCreated,
          dataKey: 'createdAt',
          className: 'createdAt',
        },
      ],
    };

    render(<Table<CellData> options={options} values={rows} />);

    screen.getByRole('columnheader', { name: headerAmount });
    screen.getByRole('columnheader', { name: headerType });
    screen.getByRole('columnheader', { name: headerCreated });
    expect(screen.getAllByRole('columnheader')).toHaveLength(3);
  });

  it('open action menu', async () => {
    interface CellData {
      id: string;
      transactionName: string;
      amount: string;
      type: string;
      createdAt: string;
    }

    const onClick = jest.fn();
    const actionText = 'Go to Details';
    const actions: ActionMenu<CellData>[] = [
      {
        id: 'actionDetails',
        type: 'default',
        value: actionText,
        action: onClick,
      },
      {
        id: 'actionDelete',
        type: 'danger',
        value: 'Delete',
      },
    ];

    const rows: CellData[] = [
      {
        id: '1',
        transactionName: 'To the Moon',
        amount: '2700',
        type: 'Back',
        createdAt: '2018-07-7 16:21:13',
      },
    ];

    const options = {
      columns: [
        {
          id: 'amount',
          value: 'Amount',
          dataKey: 'amount',
          className: 'kai',
          dataTestId: 'header-amount-id',
        },
        {
          id: 'type',
          value: 'Type',
          dataKey: 'type',
          className: 'center',
        },
        {
          id: 'created',
          value: 'Created',
          dataKey: 'createdAt',
          className: 'createdAt',
        },
        {
          id: 'transaction',
          value: 'Transaction',
          dataKey: 'transactionName',
        },
      ],
    };

    render(
      <Table<CellData> options={options} values={rows} actions={actions} />
    );

    userEvent.hover(screen.getByText(/To the Moon/i));
    userEvent.click(screen.getByTestId('icon-button'));
    expect(screen.getAllByTestId('li-action-menu')).toHaveLength(2);
    userEvent.click(screen.getByText(actionText));
    expect(onClick).toBeCalledTimes(1);
  });
});
