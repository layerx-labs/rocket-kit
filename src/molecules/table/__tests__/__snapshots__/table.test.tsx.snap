// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders 1`] = `
<DocumentFragment>
  <table
    class="sc-dmlrTW hBKmSL table"
    data-testid="table-test-id"
  >
    <thead>
      <tr>
        <th
          class="avatar"
        >
          Last Transaction
        </th>
        <th
          class="kai"
        >
          Amount
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        data-testid="row-table-test-id"
      >
        <td
          class="avatar"
          data-label="Last Transaction"
          data-testid="td-undefined"
        >
          <div>
            <img
              alt="To the Moon"
              class="sc-bdfBwQ eWkXDZ tk-avatar"
              src="./images/default-avatar.svg"
            />
             To the Moon
          </div>
        </td>
        <td
          class="kai"
          data-label="Amount"
          data-testid="td-undefined"
        >
          <div>
            2700
            <svg
              class="icon"
              viewBox="0 0 32 32"
            >
              <path
                d="M9.451 5.76c0.973 0.012 1.76 0.798 1.771 1.77v7.755l10.043-9.085c0.127-0.102 0.272-0.191 0.427-0.261l0.013-0.005c0.217-0.105 0.473-0.166 0.742-0.166 0.001 0 0.002 0 0.003 0h-0c0.492 0.012 0.934 0.214 1.258 0.536l-0-0 0.002-0.002c0.037 0.033 0.071 0.067 0.103 0.102l0.001 0.001c0.255 0.305 0.41 0.701 0.41 1.134 0 0.346-0.099 0.669-0.271 0.942l0.004-0.007c-0.053 0.076-0.107 0.143-0.166 0.205l0.001-0.001c-0.068 0.080-0.141 0.151-0.221 0.215l-0.003 0.002-7.656 6.915 1.961 1.947 1.782-1.605c0.314-0.277 0.729-0.446 1.183-0.446 0.521 0 0.989 0.222 1.317 0.576l0.001 0.001c0.273 0.296 0.44 0.693 0.44 1.129 0 0.052-0.002 0.103-0.007 0.153l0-0.006c-0.029 0.48-0.248 0.904-0.583 1.201l-0.002 0.002-1.635 1.472 3.283 3.26 0.069 0.1 0.036 0.036c0.279 0.299 0.45 0.701 0.45 1.143 0 0.026-0.001 0.052-0.002 0.077l0-0.004c-0.012 0.472-0.21 0.895-0.524 1.201l-0 0c-0.284 0.291-0.671 0.48-1.102 0.508l-0.005 0c-0.041 0.003-0.089 0.005-0.137 0.005-0.37 0-0.713-0.112-0.998-0.305l0.006 0.004-0.061-0.048-0.028-0.026c-0.032-0.022-0.010-0.007-0.066-0.047-0.028-0.022-0.055-0.047-0.102-0.092l-3.453-3.443-1.782 1.605c-0.309 0.284-0.722 0.458-1.176 0.458-0.028 0-0.057-0.001-0.085-0.002l0.004 0c-0.487-0.008-0.922-0.224-1.222-0.562l-0.002-0.002c-0.277-0.314-0.446-0.729-0.446-1.184 0-0.52 0.222-0.989 0.576-1.316l0.001-0.001 1.64-1.477-1.952-1.946-2.070 1.876v4.73c0 0.003 0 0.006 0 0.010 0 0.974-0.789 1.763-1.763 1.763-0.491 0-0.935-0.201-1.255-0.524l-0-0c-0.324-0.313-0.525-0.752-0.525-1.237 0-0.004 0-0.008 0-0.011v0.001-17.254c0.012-0.973 0.798-1.76 1.77-1.771h0.001z"
                style="fill: hsl(0, 0%, 16%);"
              />
            </svg>
          </div>
        </td>
        <td
          class="menu"
          data-testid="table-action-menu"
        >
          <div
            class="sc-crrsfI cnURua"
          >
            <button
              aria-label=""
              class="sc-gsTCUz dQVqYO button"
              color="dark"
              data-testid="icon-button"
              value=""
            >
              <svg
                class="icon"
                viewBox="0 0 32 32"
              >
                <path
                  d="M16 10.667c1.473 0 2.667-1.194 2.667-2.667s-1.194-2.667-2.667-2.667c-1.473 0-2.666 1.194-2.667 2.666v0c0.004 1.471 1.196 2.662 2.666 2.667h0zM16 13.333c-1.473 0-2.667 1.194-2.667 2.667s1.194 2.667 2.667 2.667c1.473 0 2.667-1.194 2.667-2.667v0c-0.005-1.471-1.196-2.662-2.666-2.667h-0zM16 21.333c-1.473 0-2.667 1.194-2.667 2.667s1.194 2.667 2.667 2.667c1.473 0 2.667-1.194 2.667-2.667v0c-0.005-1.471-1.196-2.662-2.666-2.667h-0z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
